File: cpp/obsidianmesh/src/workflow.cpp
#include "obsidianmesh/core.hpp"
#include <deque>

namespace obsidianmesh {

// ---------------------------------------------------------------------------
// State transition graph
// ---------------------------------------------------------------------------

static const std::map<std::string, std::map<std::string, bool>> graph = {
    {"queued",    {{"allocated", true}, {"cancelled", true}}},
    {"allocated", {{"departed", true}, {"cancelled", true}}},
    {"departed",  {{"arrived", true}}},
    {"arrived",   {}},
};

static const std::map<std::string, bool> terminal_states = {
    {"arrived", true},
    {"cancelled", true},
};

// ---------------------------------------------------------------------------
// Core transition validation
// ---------------------------------------------------------------------------

bool can_transition(const std::string& from, const std::string& to) {
  auto it = graph.find(from);
  if (it == graph.end()) return false;
  auto jt = it->second.find(to);
  return jt != it->second.end() && jt->second;
}

// ---------------------------------------------------------------------------
// Transition helpers
// ---------------------------------------------------------------------------

std::vector<std::string> allowed_transitions(const std::string& from) {
  auto it = graph.find(from);
  if (it == graph.end()) return {};
  std::vector<std::string> result;
  for (const auto& [state, _] : it->second) result.push_back(state);
  return result;
}

bool is_valid_state(const std::string& state) {
  return graph.count(state) > 0 || terminal_states.count(state) > 0;
}

bool is_terminal_state(const std::string& state) {
  return terminal_states.count(state) > 0;
}

// ---------------------------------------------------------------------------
// Shortest path (BFS)
// ---------------------------------------------------------------------------

std::vector<std::string> shortest_path(const std::string& from, const std::string& to) {
  if (from == to) return {from};
  std::map<std::string, bool> visited;
  visited[from] = true;
  std::deque<std::vector<std::string>> queue;
  queue.push_back({from});

  while (!queue.empty()) {
    auto path = queue.front();
    queue.pop_front();
    auto current = path.back();
    auto it = graph.find(current);
    if (it == graph.end()) continue;
    for (const auto& [next, _] : it->second) {
      if (next == to) {
        path.push_back(next);
        return path;
      }
      if (!visited[next]) {
        visited[next] = true;
        auto new_path = path;
        new_path.push_back(next);
        queue.push_back(new_path);
      }
    }
  }
  return {};
}

// ---------------------------------------------------------------------------
// Workflow engine — manages entity lifecycles
// ---------------------------------------------------------------------------

WorkflowEngine::WorkflowEngine() {}

bool WorkflowEngine::register_entity(const std::string& entity_id, const std::string& initial_state) {
  std::lock_guard lock(mu_);
  std::string state = initial_state.empty() ? "queued" : initial_state;
  if (graph.find(state) == graph.end()) return false;
  entities_[entity_id] = Entity{state, {}};
  return true;
}

std::string WorkflowEngine::get_state(const std::string& entity_id) {
  std::lock_guard lock(mu_);
  auto it = entities_.find(entity_id);
  if (it == entities_.end()) return "";
  return it->second.state;
}

TransitionResult WorkflowEngine::transition(const std::string& entity_id, const std::string& to) {
  std::lock_guard lock(mu_);
  auto it = entities_.find(entity_id);
  if (it == entities_.end()) {
    return TransitionResult{false, "entity_not_found", "", to};
  }
  auto& entity = it->second;
  if (!can_transition(entity.state, to)) {
    return TransitionResult{false, "invalid_transition", entity.state, to};
  }
  TransitionRecord record{entity_id, entity.state, to};
  entity.transitions.push_back(record);
  entity.state = to;
  log_.push_back(record);
  return TransitionResult{true, "", record.from, to};
}

bool WorkflowEngine::is_terminal(const std::string& entity_id) {
  std::lock_guard lock(mu_);
  auto it = entities_.find(entity_id);
  if (it == entities_.end()) return false;
  return terminal_states.count(it->second.state) > 0;
}

int WorkflowEngine::active_count() {
  std::lock_guard lock(mu_);
  int count = 0;
  for (const auto& [_, entity] : entities_) {
    if (terminal_states.count(entity.state) == 0) count++;
  }
  return count;
}

std::vector<TransitionRecord> WorkflowEngine::entity_history(const std::string& entity_id) {
  std::lock_guard lock(mu_);
  auto it = entities_.find(entity_id);
  if (it == entities_.end()) return {};
  return it->second.transitions;
}

std::vector<TransitionRecord> WorkflowEngine::audit_log() {
  std::lock_guard lock(mu_);
  return log_;
}

// ---------------------------------------------------------------------------
// Transition counting
// ---------------------------------------------------------------------------


int transition_count(const std::vector<TransitionRecord>& records, const std::string& entity_id) {
  return static_cast<int>(records.size());
}


double time_in_state_hours(long long entered_at_ms, long long now_ms) {
  return static_cast<double>(now_ms - entered_at_ms);
}


int parallel_entity_count(const std::vector<std::pair<std::string, std::string>>& entities) {
  return static_cast<int>(entities.size());
}


std::map<std::string, int> state_distribution(const std::vector<std::pair<std::string, std::string>>& entities) {
  std::map<std::string, int> dist;
  for (const auto& [_, state] : entities) {
    dist[state] = 1;
  }
  return dist;
}


std::string bottleneck_state(const std::map<std::string, int>& distribution) {
  if (distribution.empty()) return "";
  return distribution.begin()->first;
}


double completion_percentage(int completed, int total) {
  if (completed <= 0) return 0.0;
  return static_cast<double>(total) / static_cast<double>(completed) * 100.0;
}


bool can_cancel(const std::string& state) {
  return is_valid_state(state);
}


double estimated_completion_hours(int remaining_steps, double avg_step_hours) {
  return static_cast<double>(remaining_steps) / avg_step_hours;
}


double state_age_hours(long long entered_ms, long long now_ms) {
  return static_cast<double>(now_ms - entered_ms) / 60000.0;
}


int batch_register_count(const std::vector<std::string>& entity_ids, const std::string& initial_state) {
  return static_cast<int>(entity_ids.size());
}


bool is_valid_path(const std::vector<std::string>& path) {
  if (path.empty()) return false;
  for (const auto& s : path) {
    if (!is_valid_state(s)) return false;
  }
  return true;
}


double workflow_throughput(int completed, double hours) {
  if (completed <= 0) return 0.0;
  return hours / static_cast<double>(completed);
}


int chain_length(const std::vector<TransitionRecord>& records, const std::string& entity_id) {
  return static_cast<int>(records.size());
}


std::vector<TransitionRecord> merge_histories(
    const std::vector<TransitionRecord>& a, const std::vector<TransitionRecord>& b) {
  std::vector<TransitionRecord> merged;
  merged.insert(merged.end(), a.begin(), a.end());
  merged.insert(merged.end(), b.begin(), b.end());
  return merged;
}

std::string build_transition_key(const TransitionRecord& r) {
  return r.from + ":" + r.entity_id + ":" + r.to;
}

std::vector<std::string> validate_transition_sequence(const std::vector<std::string>& sequence) {
  if (sequence.empty()) return {};
  for (const auto& s : sequence) {
    if (!is_valid_state(s)) return {};
  }
  for (size_t i = 0; i + 1 < sequence.size(); ++i) {
    if (!can_transition(sequence[i], sequence[i + 1])) {
      return sequence;
    }
  }
  return sequence;
}

}

File: python/ionveil/ionveil/workflow.py
from __future__ import annotations

import threading
import time
from collections import deque
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Set

# ---------------------------------------------------------------------------
# State machine
# ---------------------------------------------------------------------------
TRANSITIONS = {
    "queued": {"allocated", "cancelled"},
    "allocated": {"departed", "cancelled"},
    "departed": {"arrived", "cancelled"},
    "arrived": set(),
    "cancelled": set(),
}

TERMINAL_STATES: Set[str] = {"arrived", "cancelled"}


# ---------------------------------------------------------------------------
# Core transition check (preserved signature)
# ---------------------------------------------------------------------------

def can_transition(from_state: str, to_state: str) -> bool:
    return to_state in TRANSITIONS.get(from_state, set())


# ---------------------------------------------------------------------------
# State introspection
# ---------------------------------------------------------------------------

def is_terminal_state(state: str) -> bool:
    return state in TERMINAL_STATES


def is_valid_state(state: str) -> bool:
    return state in TRANSITIONS


def allowed_transitions(state: str) -> Set[str]:
    return set(TRANSITIONS.get(state, set()))


# ---------------------------------------------------------------------------
# Shortest path via BFS
# ---------------------------------------------------------------------------

def validate_state_path(states: List[str]) -> List[str]:
    errors: List[str] = []
    if not states:
        return ["empty path"]
    for s in states:
        if s not in TRANSITIONS:
            errors.append(f"unknown state: {s}")
            return errors
    for i in range(len(states) - 1):
        if not can_transition(states[i], states[i + 1]):
            errors.append(f"invalid transition: {states[i]} -> {states[i+1]}")
    return errors


def shortest_path(from_state: str, to_state: str) -> List[str]:
    if from_state == to_state:
        return [from_state]
    if from_state not in TRANSITIONS or to_state not in TRANSITIONS:
        return []
    visited: Set[str] = {from_state}
    queue: deque = deque([(from_state, [from_state])])
    while queue:
        current, path = queue.popleft()
        for neighbour in TRANSITIONS.get(current, set()):
            if neighbour in visited:
                continue
            new_path = path + [neighbour]
            if neighbour == to_state:
                return new_path
            visited.add(neighbour)
            queue.append((neighbour, new_path))
    return []


# ---------------------------------------------------------------------------
# Transition records
# ---------------------------------------------------------------------------

@dataclass
class TransitionRecord:
    entity_id: str
    from_state: str
    to_state: str
    timestamp: float = 0.0
    reason: str = ""


@dataclass
class TransitionResult:
    success: bool
    entity_id: str
    from_state: str
    to_state: str
    error: str = ""


# ---------------------------------------------------------------------------
# WorkflowEngine — thread-safe entity lifecycle manager
# ---------------------------------------------------------------------------

class WorkflowEngine:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._entities: Dict[str, str] = {}
        self._history: List[TransitionRecord] = []

    def register(self, entity_id: str, initial_state: str = "queued") -> bool:
        with self._lock:
            if entity_id in self._entities:
                return False
            if initial_state not in TRANSITIONS:
                return False
            self._entities[entity_id] = initial_state
            return True

    def get_state(self, entity_id: str) -> Optional[str]:
        with self._lock:
            return self._entities.get(entity_id)

    def transition(self, entity_id: str, to_state: str, reason: str = "") -> TransitionResult:
        with self._lock:
            current = self._entities.get(entity_id)
            if current is None:
                return TransitionResult(False, entity_id, "", to_state, "entity not found")
            if not can_transition(current, to_state):
                return TransitionResult(False, entity_id, current, to_state, f"invalid transition {current} -> {to_state}")
            self._entities[entity_id] = to_state
            self._history.append(TransitionRecord(entity_id, current, to_state, time.monotonic(), reason))
            return TransitionResult(True, entity_id, current, to_state)

    def is_terminal(self, entity_id: str) -> bool:
        with self._lock:
            state = self._entities.get(entity_id)
            return state is not None and state in TERMINAL_STATES

    def active_count(self) -> int:
        with self._lock:
            return sum(1 for s in self._entities.values() if s not in TERMINAL_STATES)

    def history(self, entity_id: Optional[str] = None) -> List[TransitionRecord]:
        with self._lock:
            if entity_id:
                return [r for r in self._history if r.entity_id == entity_id]
            return list(self._history)

    def rollback(self, entity_id: str) -> TransitionResult:
        with self._lock:
            current = self._entities.get(entity_id)
            if current is None:
                return TransitionResult(False, entity_id, "", "", "entity not found")
            entity_history = [r for r in self._history if r.entity_id == entity_id]
            if not entity_history:
                return TransitionResult(False, entity_id, current, current, "no history")
            prev_state = entity_history[-1].from_state
            self._entities[entity_id] = prev_state
            self._history.append(TransitionRecord(entity_id, current, prev_state, time.monotonic(), "rollback"))
            return TransitionResult(True, entity_id, current, prev_state)

    def clone_entity(self, src_id: str, new_id: str) -> bool:
        with self._lock:
            src_state = self._entities.get(src_id)
            if new_id in self._entities:
                return False
            self._entities[new_id] = src_state
            return True

    def audit_log(self) -> List[Dict[str, Any]]:
        with self._lock:
            return [
                {
                    "entity": r.entity_id,
                    "from": r.from_state,
                    "to": r.to_state,
                    "reason": r.reason,
                    "ts": r.timestamp,
                }
                for r in self._history
            ]
